[JENKINS-56005] Add user property-change listener

davidolorundare opened the pull request:
See [JENKINS-56005](https://issues.jenkins-ci.org/browse/JENKINS-56005) and [JENKINS-56008](https://issues.jenkins-ci.org/browse/JENKINS-56008).

- Revisions based on the discussions in the comments, an interface; `UserPropertyListener` was created, with `default` method all other listener implementations would make use of. Enabling support for listening to Jenkins user property-change events. 
- Two new listener implementation classes; `PasswordPropertyListener` and `ApiTokenPropertyListener` were created and added to jenkins-core. These classes handle firing notifications specific to their respective use cases (i.e. password-change and api-token creation/revocation events respectively). The rationale for the additions is discussed in each above ticket.
- The relevant methods, in the corresponding jenkins-core classes, where these events occur; `HudsonPrivateSecurityRealm.Details` and `ApiTokenProperty`, were updated with calls to the respective new listener class methods.
- The corresponding testsuites for each; `HudsonPrivateSecurityRealmTest` and `ApiTokenPropertyTest` were updated with tests that exercise these new listener class methods.

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Enhancement: Addition of new `hudson.model.UserPropertyListener`interface to support notification of current and future Jenkins user-account property-change events (e.g. password changes, api-token key creation and deletion, etc).
* Enhancement: Addition of new `hudson.security.PasswordPropertyListener`class to support notification of Jenkins user-account password property-change events.
* Enhancement: Addition of new `jenkins.security.ApiTokenPropertyListener`class to support notification of Jenkins user-account api-token creation and revocation events.
* Enhancement: Update to `hudson.security.HudsonPrivateSecurityRealm.Details` adding notification(s) of user-account password property-change events to its method.
* Enhancement: Update to `jenkins.security.ApiTokenProperty` adding notification(s) of user-account api-token creation and revocation events to its method.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers
@jvz 
@jeffret-b 
@Wadeck 
@daniel-beck 


<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

davidolorundare committed:
[JENKINS-56005] Add user property-change listener

davidolorundare committed:
[JENKINS-56008] Update security-realm with listener

davidolorundare committed:
[JENKINS-56008] Add user property-change unit test

Added a new unit test which evaluates logging of a
user property-change (update of user's password).

davidolorundare commented:
restarting build

jvz commented on core/src/main/java/jenkins/security/SecurityListener.java:
```diff
@@ -78,6 +79,16 @@ protected void loggedIn(@Nonnull String username){}
+     * @param p the user object property
+     * @param username the username
+     */
+    protected void userPropertyUpdated(@Nonnull UserProperty p, @Nonnull String username){}
```
I kind of wonder if this would be more appropriate in its own `UserPropertyListener` or similar. I'm thinking in terms that it may eventually be useful to have event listeners for JobProperty, NodeProperty, and others if I'm forgetting.

Also, I'm not sure if there's precedence in this class already, but I think it might make more sense to flip the ordering of the parameters here.

jeffret-b replied:
I'm not really convinced one way or another. The `userPropertyUpdated()` method doesn't quite seem to fit the other things in SecurityListener. But then, some of the other items in SecurityListener don't fit well either. The user property is more about users than specifically about security. But so are some of the other things in this file.

I guess when it comes down to it I have a slight preference for separating out a `UserPropertyListener`.

davidolorundare replied:
hmh.. alright. Question though is for the new `UserPropertyListener`: it being an abstract-class versus interface ?

davidolorundare replied:
@Wadeck Any thoughts on the creation of a new separate `UserPropertyListener` ?

Wadeck replied:
Separation seems better for the long term, like Matt mentioned, about the other items that could/would require such listeners in the future. And also, it's not directly security related, so not really a reason to be put in `SecurityListener`, except for the specific case of `Details`. But I imagine the goal of this listener is to be triggered everytime a property change, in general.

davidolorundare replied:
@Wadeck abstract versus interface ?

Wadeck replied:
I will bet on `abstract` being the most natural way as you are providing some static helper methods (for the event trigger). The interfaces should be used for contract definition.

jvz replied:
Interfaces allow static methods as of Java 8. See for example the methods added to the collections API. Also default methods.

jvz commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -342,6 +346,41 @@ public void userCreationWithHashedPasswords() throws Exception {
+    @Issue("JENKINS-56008")
+    @Test
+    public void updateUserPassword() throws Exception {
+        HudsonPrivateSecurityRealm securityRealm = new HudsonPrivateSecurityRealm(true, false, null);
```
Not sure if you've tried these before, but perhaps you can simplify these tests a bit by using `JenkinsRule.createDummySecurityRealm()`? I haven't tried it myself, though I have seen many tests use it.

davidolorundare replied:
hmh.. I've tried it before in some previous tests, but wasn't sure how well it mimics a real user account compared to creating one manually. Will take another look at it though.

davidolorundare replied:
I looked into this, reference here - https://gitter.im/jenkinsci/outreachy?at=5c6c4c58ab952d30855a538e

jvz commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -342,6 +346,41 @@ public void userCreationWithHashedPasswords() throws Exception {
+        signup.enterUsername("debbie");
+        signup.enterPassword("debbie");
+        signup.enterFullName(StringUtils.capitalize("debbie user"));
+        signup.enterEmail("debbie" + "@" + "debbie" + ".com");
```
I like to use example.com as an example URL because it'll always be a sample domain. See http://example.com/

davidolorundare replied:
Will add this in the next commit.

jvz commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -342,6 +346,41 @@ public void userCreationWithHashedPasswords() throws Exception {
+
+        WebRequest request = new WebRequest(configPage, HttpMethod.POST);
+        request.setAdditionalHeader("Content-Type", "application/x-www-form-urlencoded");
+        request.setRequestBody("json=" + URLEncoder.encode(formData, StandardCharsets.UTF_8.name()));
```
I can't believe this is how form encoding actually works. Ugh. :/

davidolorundare replied:
lol

jeffret-b reviewed (commented):
It's looking good.

Wadeck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -717,14 +717,18 @@ public Details newInstance(StaplerRequest req, JSONObject formData) throws FormE
                 if(data!=null) {
                     String prefix = Stapler.getCurrentRequest().getSession().getId() + ':';
                     if(data.startsWith(prefix))
+                        SecurityListener.fireUserPropertyUpdated(p, user.getId());
```
⚠️ missing `{}` in the `if`, so you are breaking the statement there.

That's exactly the main reason why I *hate* the `if` without brackets 👿 

If that code bug was not discovered by your tests => there is a missing path in your coverage.

davidolorundare replied:
do you advise using the `if` for only one-liners; or to just never use a non-bracketed `if` at all ?

Like;
```
                if((data!=null) && (p != null)) {
                    String prefix = Stapler.getCurrentRequest().getSession().getId() + ':';
                    if(data.startsWith(prefix)) return Details.fromHashedPassword(data.substring(prefix.length()));
                }
```

versus;

```                if((data!=null) && (p != null)) {
                    String prefix = Stapler.getCurrentRequest().getSession().getId() + ':';
                    if(data.startsWith(prefix)) {
                        return Details.fromHashedPassword(data.substring(prefix.length()));
                    }
                }
```

Wadeck replied:
I can only advice to never use a `if` without the associated `{}`, to avoid situation where someone wants to add a line, but due to the missing brackets will break the code flow like here.

And the one-liner is worse from my PoV due to the poor readability.

davidolorundare replied:
perks of using Python over Java :)

jvz replied:
I only use one line ifs for things like:

```if (foo == null) return null;```

or

```if (bar) return false;```

or

```if (baz) return;```

Basically, only for control flow type things. Remember, forgetting your brackets can cause security problems (see the old Apple bug). If you auto-format your code before committing, you can see if you missed brackets based on how it indents things.

Wadeck commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -717,14 +717,18 @@ public Details newInstance(StaplerRequest req, JSONObject formData) throws FormE
                     throw new FormException("Please confirm the password by typing it twice","user.password2");
 
+                User user = Util.getNearestAncestorOfTypeOrThrow(req, User.class);
+                UserProperty p = user.getProperty(Details.class);
```
Did you check **all** the paths that lead to a User and ensured they all have the property set? Example: after changing security realm.

My concern is the `getProperty` could return `null` but you explicitly say you want a `@Nonnull` property and there is no check around that in your code.

davidolorundare replied:
good point! updated in the latest commit.

Wadeck commented:
Is such approach really required? I mean to implement event firing specifically per property? I hope there is a better way to do that globally, without modification of plugins, like in `UserProperty#reconfigure`. Not sure it's feasible but that should be considered first.

Without a global one-time modification, you have to wait for all the plugins to implements that new event trigger, before being able to rely on it.

In addition, the current approach is the opposite of DRY.

Another point, what happen if I save a user, without modification? From even listener PoV, the property was not updated, but in reality it will be recreated. That's something to take into account, ideally by not triggering the event if the property is not **really** changed, otherwise it will be the responsibility of the listener to check such things and that could become really cumbersome.

Finally, the new value is not really sufficient / useful if we do not have the previous value of the property. I imagine more an event with the previous value (potentially null in case of creation of the property), the new value (potentially null in case of removal of that property) and the User as it's not retrievable from the property directly.

jvz commented:
> Is such approach really required? I mean to implement event firing specifically per property? I hope there is a better way to do that globally, without modification of plugins, like in `UserProperty#reconfigure`. Not sure it's feasible but that should be considered first.

It could be useful to support audit logging changes to your API tokens, credentials (user creds at least; node creds need a separate NodeProperty listener), maybe even some other details. It would be nice for the UserProperty class itself to automatically fire the events rather than consumers.

> Another point, what happen if I save a user, without modification? From even listener PoV, the property was not updated, but in reality it will be recreated. That's something to take into account, ideally by not triggering the event if the property is not **really** changed, otherwise it will be the responsibility of the listener to check such things and that could become really cumbersome.

Agreed. We should track the old and new values. These can be compared with `.equals()` or whatever appropriate to see if it changed, and then that would fire the event. In that case, it may be useful to include both properties as parameters (old and new).

davidolorundare committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

davidolorundare committed:
Refactor securitylistener methods

davidolorundare committed:
[JENKINS-56005] Add userpropertylistener interface

davidolorundare committed:
[JENKINS-56008] Add passwordpropertylistener class

davidolorundare committed:
[JENKINS-56008] Refactor private-security realm

Updated the HudsonPrivateSecurityRealm newInstance()
method to use the PasswordPropertyListener class for
notifying subscribers on changes to a user's password.

davidolorundare committed:
[JENKINS-56170] Add apitokenpropertylistener class

davidolorundare committed:
[JENKINS-56170] Update apitokenproperty methods

Updated the ApiTokenProperty doGenerateToken() and
doRevoke() methods to use the ApiTokenPropertyListener
class for notifying subscribers of api-token creations
and revocations from the Jenkins UI, respectively.

davidolorundare committed:
[JENKINS-56008] Update private-security realm unittest

Updated HudsonPrivateSecurityRealmTest with a unittest
that exercises the added PasswordPropertyListener class
for notifying subscribers of user password changes.

davidolorundare committed:
[JENKINS-56170] Update apitokenproperty unittests

Added two unittests to the ApiTokenPropertyTest class
for exercising the ApiTokenPropertyListener class methods;
for notifying subscribers of api-token creation and revocation
events.

davidolorundare commented:
some updates have been made to the PR based on the earlier discussions.
cc @jeffret-b @jvz @Wadeck @daniel-beck

jvz commented on core/src/main/java/hudson/model/UserPropertyListener.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     */
+    default void onChanged(@Nonnull String username, @Nonnull Object oldValue, @Nonnull Object newValue) {
+        LOGGER.log(Level.FINE, MessageFormat.format("{0} property changed for user {1}", oldValue.toString(), username));
```
Is this a duplicate of the other onChanged method?

jvz reviewed (commented):
I like where this is going, but I'm not clear on the multiple signatures for the same events.

CC: @jenkinsci/code-reviewers for some API feedback.

davidolorundare committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins

davidolorundare committed:
Refactor userpropertylistener methods

davidolorundare commented:
Restarting the build for this pull request.

varyvol commented:
@davidolorundare did you really intend to close this or was it a mistake?

davidolorundare commented:
Hi @varyvol, I'm going to reopen it soon; I need it to rebuild again
> @davidolorundare did you really intend to close this or was it a mistake?

jvz commented on core/src/main/java/hudson/model/UserPropertyListener.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     */
+    default void onCreated(@Nonnull String username, @Nonnull UserProperty value) {
+        LOGGER.log(Level.FINE, MessageFormat.format("new {0} property created for user {1}", value.getClass().toString(), username));
```
My suggestion on syntax when using `java.util.logging` is to take advantage of the lambda forms and use normal string concatenation. In this line, that would be:

```suggestion
        LOGGER.fine(() -> "new " + value.getClass().toString() + " property created for user " + username));
```

jvz commented on core/src/main/java/hudson/model/UserPropertyListener.java:
```diff
@@ -0,0 +1,111 @@
+     * @param newValue new property of the user
+     *
+     */
+    default <T> void onChanged(@Nonnull String username, @Nonnull T oldValue, @Nonnull T newValue) {
```
Same issue as `<T>onCreated` here.

jvz commented on core/src/main/java/hudson/model/UserPropertyListener.java:
```diff
@@ -0,0 +1,111 @@
+     * @param value property that was removed
+     *
+     */
+    default <T> void onDeleted(@Nonnull String username, @Nonnull T value) {
```
Same issue as `<T>onCreated` here.

jvz commented on core/src/main/java/hudson/security/PasswordPropertyListener.java:
```diff
@@ -0,0 +1,59 @@
+     * @param newValue new password of the user
+     *
+     * **/
+    static void fireOnChanged(@Nonnull String username, @Nonnull String oldValue, @Nonnull String newValue) {
```
You should probably note here somehow that the values are hashed passwords, not the actual password values. Otherwise, I was about to suggest that this method should use `char[]` instead of `String` so that you could zero out the arrays after using them, but that's overkill for hashes.

jvz commented on core/src/main/java/hudson/model/UserPropertyListener.java:
```diff
@@ -0,0 +1,111 @@
+/**
+ * Listener interface which all other user property-specific event-listeners make use of.
+ */
+public interface UserPropertyListener extends ExtensionPoint {
```
If you made this `UserPropertyListener<T extends UserProperty> extends ExtensionPoint` and then used `T` instead of `UserProperty` below, you'd be able to better express that a listener is for a specific property type. Then the static methods can look up the appropriate listener based on the property type. That would work better than the existing parameterization I think.

varyvol commented on core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:
```diff
@@ -717,14 +717,21 @@ public Details newInstance(StaplerRequest req, JSONObject formData) throws FormE
                 }
 
-                User user = Util.getNearestAncestorOfTypeOrThrow(req, User.class);
+                if (p != null) {
```
You are checking here if p is not null, but you already called the getPassword() method a few lines before.

rsandell commented on core/src/main/java/hudson/model/UserPropertyListener.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     */
+    default void onChanged(@Nonnull String username, @Nonnull UserProperty oldValue, @Nonnull UserProperty newValue) {
+        LOGGER.log(Level.FINE, MessageFormat.format("{0} property changed for user {1}", oldValue.getClass().toString(), username));
```
Having the default implementation log will mean that every single listener implementation that doesn't implement the method will spam the log.

varyvol commented on core/src/main/java/hudson/security/PasswordPropertyListener.java:
```diff
@@ -0,0 +1,59 @@
+ * Listener notified of user password change events from the jenkins UI.
+ */
+@Extension
+public class PasswordPropertyListener implements UserPropertyListener {
```
Which is the use case that needs receiving the password (even if hashed)? Detecting the change is fine, but this could allow retrieving the actual values.

varyvol reviewed (commented):
@davidolorundare I made a couple suggestions and I also agree with @jvz 's comments. Do you think you could go through the feedback?

jglick commented:
Closing for now as this seems to have been abandoned.

Normally we would add the `stalled-pr` label to the corresponding JIRA issue, to make it easier to search for patches that someone could pick up and revive, but I cannot find one in this case—the two linked issues are in the `audit-log-plugin` component, and marked fixed.

zbynek commented:
@jglick issues seem to be moved to GitHub https://github.com/jenkinsci/audit-log-plugin/issues/37
